Try: High-luminance default colors

jasmussen opened the pull request:
These are colors that are explored through the G2 visual refresh (see #18667).

<img width="1747" alt="Screenshot 2020-05-11 at 12 30 00" src="https://user-images.githubusercontent.com/1204802/81552138-348ee400-9383-11ea-99a9-731376132c2f.png">

The color change does a few things:

- It is chosen and balanced against the rest of the visual refresh.
- It reduces the overall color palette, removing secondary colors used for toggles or checkmarks, unifying on a single blue.
- It increases contrast against white, quite a bit, which is good for toggles, focus, form widgets and buttons.

The new blue has a contrast of 5.61:1 compared to 4.57:1 for the previously used WordPress blue (#007cba).

Before:

<img width="735" alt="Screenshot 2020-02-26 at 12 11 47" src="https://user-images.githubusercontent.com/1204802/75339790-784c7400-5891-11ea-8bcc-07c9c41e09ca.png">

After:

<img width="740" alt="Screenshot 2020-02-26 at 12 09 45" src="https://user-images.githubusercontent.com/1204802/75339802-7b476480-5891-11ea-88c0-720cba38eb1f.png">

Commit:
Try: High-luminance default colors

These are colors that are explored through the G2 visual refresh (see #18667).

The color change does a few things:

- It is chosen and balanced against the rest of the visual refresh.
- It reduces the overall color palette, removing secondary colors used for toggles or checkmarks, unifying on a single blue.
- It increases contrast against white, quite a bit, which is good for toggles, focus, form widgets and buttons.

The new blue has a contrast of 5.61:1 compared to 4.57:1 for the previously used WordPress blue (#007cba).

youknowriad commented on packages/base-styles/_colors.scss:
```diff
@@ -31,7 +31,7 @@ $medium-gray-text: #757575;
 $light-gray-tertiary: #e7e8e9;
 $theme-color: theme(button);
-$blue-medium-focus: #007cba;	// @todo: Currently being used as "spot" color. Needs to be considered in context of themes.
+$blue-medium-focus: #3858e9;	// @todo: Currently being used as "spot" color. Needs to be considered in context of themes.
```
Let's try to find a way to compute this based on "theme" (otherwise the alternative schemes are a bit broken)

jasmussen replied:
I pushed a fix. There really aren't that many shades of blue used here, but what shades there are should take from this now, right?

youknowriad reviewed (commented):
What's the plan for WordPress Core here? I'm assuming we want to stay consistent. Should we create a trac ticket, make/core post for the next steps?

ellatrix commented:
😍 I like this a lot.

mtias commented:
> What's the plan for WordPress Core here? I'm assuming we want to stay consistent. Should we create a trac ticket, make/core post for the next steps?

I think core should overwrite the default one initially.

SchneiderSam commented:
It looks fresher. I like it.  And yes...please be consistent with core... 😍

Commit:
Update to use theme color.

jasmussen commented:
I updated to use theme colors, but kept the update to the default theme:

<img width="1161" alt="Screenshot 2020-02-27 at 11 18 02" src="https://user-images.githubusercontent.com/1204802/75435183-f6248400-5952-11ea-9966-ad4c755cc0ba.png">

Here's Sunrise:

<img width="1161" alt="Screenshot 2020-02-27 at 11 17 39" src="https://user-images.githubusercontent.com/1204802/75435192-fa50a180-5952-11ea-9694-b4bf827d7597.png">

Here's Midnight:

<img width="1159" alt="Screenshot 2020-02-27 at 11 17 50" src="https://user-images.githubusercontent.com/1204802/75435205-00468280-5953-11ea-9de1-3f94f32a8953.png">

This brings an excellent segue to the conversation on the plan for the core project. Honestly, every color theme here deserves an update, not just the default one. The Midnight one is lovely, but the idea of using the red spot color for buttons is perhaps not the best, as it indicates "danger". That aspect is unchanged, but worth bringing up.

Matías plan makes sense:

> I think core should overwrite the default one initially.

That is, we can move forward with the fresh blue for the Gutenberg project itself, but we create a Trac ticket to add CSS to the core project to override these colors back to the #007bca color that the rest of the admin interface ships with. This keeps things consistent, even if it does mean anyone using the block editor as part of WordPress will get the lower-luminance colors, and even if that means the majority of users won't see these blues in the near future. 

It seems like a trac ticket to suggest that is the next step here?

youknowriad commented:
> That is, we can move forward with the fresh blue for the Gutenberg project itself, but we create a Trac ticket to add CSS to the core project to override these colors back to the #007bca color that the rest of the admin interface ships with

I don't think it's a great idea personally, because that variable usage can change here. We can add new components using this variable or update existings ones but we won't think that we need to do a trac ticket for each one of these changes to update the override.

jasmussen commented:
> I don't think it's a great idea personally, because that variable usage can change here. We can add new components using this variable or update existings ones but we won't think that we need to do a trac ticket for each one of these changes to update the override.

@youknowriad Here's an idea:

> When we build the plugin we replace the color

Can you help me do that? 

As it is now, there's a primary and a secondary color, that's all that's left, and the rewriting would need to do this:

#3858e9 → #007cba
#1635bb → #11a0d2

But we could even simplify the base-styles further, as you can see the G2 UI reduces the need for many colors. We could make it a single spot color that's used everywhere, so all we'd have to do is this:

#3858e9 → #007cba

Let me know what you think.

youknowriad commented:
@jasmussen When Core consumes the styles, they are already built into CSS, so it's not just a single place or two, it's everywhere the variable is used and the only way for Core to override is to add a stylesheet overriding all the places where the variable was used.

mtias commented:
This would be before core. A different build for the playground and the plugin / core using a different base color.

youknowriad commented:
> A different build for the playground and the plugin / core using a different base color.

Yes,  this is potentially the solution. Have two separate CSS outputs for the packages. I wonder about the performance impact on the build time.

mtias commented:
Would it make sense to run it only when creating the plugin bundle? (Then master would get the updated blue, but the plugin would get the WordPress blue.)

jasmussen commented:
Rebased this to just freshen it up:

<img width="1310" alt="Screenshot 2020-03-10 at 11 05 41" src="https://user-images.githubusercontent.com/1204802/76301780-25c27d00-62bf-11ea-9e5a-3bc36e9f3c77.png">

I could use help with the next steps.

jasmussen commented:
Rebased this and updated the screenshot in the PR ticket just as a refresher.

I love the new color, I think it would be a great default for contexts outside of WordPress.

But because it needs dev help to make that happen, I'm guessing this PR will go stale eventually, and so I extracted a small change into https://github.com/WordPress/gutenberg/pull/22261 which can easily go in soon.

youknowriad commented:
cc @gziolo since you were working on the CSS support for scripts. 

For context: we need a way to generate the stylesheets for two color schemes. One style to be used on "Core" with the default core colors (existing colors), and one to be used in the plugin with the high-luminance default colors.

gziolo commented:
> cc @gziolo since you were working on the CSS support for scripts.
> 
> For context: we need a way to generate the stylesheets for two color schemes. One style to be used on "Core" with the default core colors (existing colors), and one to be used in the plugin with the high-luminance default colors.

Does it mean, we want a different config depending on the context, i.e:
1. One set of colors for that lands in packages and is consumed by WordPress core, 3rd party projects, or Storybook.
2. The alternative set is used only by the Gutenberg plugin.

Do I get it correct?

youknowriad commented:
@gziolo yes, I guess dotcom might also want to opt-in into the "plugin's" set

gziolo commented:
Does it have to work like this forever or it's only until the plugin and WP core align? It's important because we could use `process.env.WP_HIGH_LUMINANCE` flag to control it if it's temporary :)

youknowriad commented:
I'm not sure how realistic is it to expect Core to use these colors everywhere anytime soon. That said, the requirement is that npm consumers could choose one or the other which means an env variable won't work, we'd need to ship both.

gziolo commented:
It feels like we could add a new section to `@wordpress/base-styles` next to `default`. You should know better how much work it would be to update postcss plugin to consume this section. I think you authored this plugin, didn't you?

ellatrix commented:
I'd love this to happen. Is there anyone who could help out?

jasmussen commented:
> I'd love this to happen. Is there anyone who could help out?

In https://github.com/WordPress/gutenberg/pull/22727#discussion_r432285474 we discussed how this PR might not be the approach we want to take after all.

The intent of this PR is to bake in the high luminance blue color by default, so environments that use the block editor outside of wp-admin inherit this color. But wp-admin should still override it with existing brand colors. 

A slightly different approach would be to allow the environment that includs the block editor to provide this color, rather than rely on a default (which there might still be). This might involve some PostCSS magic that @gziolo has been thinking about. Ideally this should work in a way that is more conducive to theming — not with a ton of colors, maybe just a single spot color.

youknowriad commented:
> A slightly different approach would be to allow the environment that includs the block editor to provide this color, rather than rely on a default (which there might still be). This might involve some PostCSS magic that @gziolo has been thinking about. Ideally this should work in a way that is more conducive to theming — not with a ton of colors, maybe just a single spot color.

Assuming it's fine to use CSS variables, the simplest solution here would be to use a CSS variable and change its value depending on the context. I know the #core-css folks on Core are experimenting with CSS variables for admin theme colors which would solve this issue.

jasmussen commented:
I'd love to move to a CSS variable.

mtias commented:
How about creating a new separate color scheme for the admin?

jasmussen commented:
> How about creating a new separate color scheme for the admin?

I.e. new blue by default but overridden by an admin color theme? That could work too.

ellatrix commented:
Yes, it would be great to even have a new default and leave the rest alone for now.

jasmussen commented:
Closing in favor of #23048!
